paddle/framework/scope.h

```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "paddle/framework/variable.h"
+#include "paddle/platform/assert.h"
+
+namespace paddle {
+namespace framework {
+
+/**
+ * Scope is an association of a name to Variable. All variables belong to
+ * Scope. You need to specify a scope to run a Net, i.e., `net.Run(&scope)`.
+ * One net can run in different scopes and update different variable in the
+ * scope.
+ */
+class Scope {
+ private:
+  explicit Scope(const std::shared_ptr<Scope>& parent = nullptr)
+      : parent_(parent) {}
+
+ public:
+  static std::shared_ptr<Scope> Create(
+      const std::shared_ptr<Scope>& parent = nullptr) {
+    return std::make_shared<Scope>(Scope(parent));
```

Comment:
it will fail if the default consturctor is private:
```shell
In file included from /Users/qiaolongfei/project/paddle/paddle/framework/scope_test.cc:15:
In file included from /Users/qiaolongfei/project/paddle/paddle/framework/scope.h:17:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/string:439:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:628:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2194:15: error: field of type 'paddle::framework::Scope' has private default constructor
              __second_(_VSTD::forward<_Args2>(_VSTD::get<_I2>(__second_args))...)
              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2457:15: note: in instantiation of function template specialization 'std::__1::__libcpp_compressed_pair_imp<std::__1::allocator<paddle::framework::Scope>, paddle::framework::Scope, 1>::__libcpp_compressed_pair_imp<std::__1::allocator<paddle::framework::Scope> &, const std::__1::shared_ptr<paddle::framework::Scope> &, 0, 0>' requested here
            : base(__pc, _VSTD::move(__first_args), _VSTD::move(__second_args),
              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3800:16: note: in instantiation of function template specialization 'std::__1::__compressed_pair<std::__1::allocator<paddle::framework::Scope>, paddle::framework::Scope>::__compressed_pair<std::__1::allocator<paddle::framework::Scope> &, const std::__1::shared_ptr<paddle::framework::Scope> &>' requested here
            :  __data_(piecewise_construct, _VSTD::forward_as_tuple(__a),
               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:4411:26: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<paddle::framework::Scope, std::__1::allocator<paddle::framework::Scope> >::__shared_ptr_emplace<const std::__1::shared_ptr<paddle::framework::Scope> &>' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:4775:29: note: in instantiation of function template specialization 'std::__1::shared_ptr<paddle::framework::Scope>::make_shared<const std::__1::shared_ptr<paddle::framework::Scope> &>' requested here
    return shared_ptr<_Tp>::make_shared(_VSTD::forward<_Args>(__args)...);
                            ^
/Users/qiaolongfei/project/paddle/paddle/framework/scope.h:41:17: note: in instantiation of function template specialization 'std::__1::make_shared<paddle::framework::Scope, const std::__1::shared_ptr<paddle::framework::Scope> &>' requested here
    return std::make_shared<Scope>(parent);
                ^
/Users/qiaolongfei/project/paddle/paddle/framework/scope.h:35:12: note: declared private here
  explicit Scope(const std::shared_ptr<Scope>& parent = nullptr)
           ^
1 error generated.
make[2]: *** [paddle/framework/CMakeFiles/scope_test.dir/scope_test.cc.o] Error 1
make[1]: *** [paddle/framework/CMakeFiles/scope_test.dir/all] Error 2
```